[visionOS] Update OverlayRegion configuration more deliberately

etiennesegonzac opened the pull request:
#### 42cf0519aef9309ac80daefc2f6b30d9af0e3308
<pre>
[visionOS] Update OverlayRegion configuration more deliberately
<a href="https://bugs.webkit.org/show_bug.cgi?id=303394">https://bugs.webkit.org/show_bug.cgi?id=303394</a>
&lt;<a href="https://rdar.apple.com/163802576">rdar://163802576</a>&gt;

Reviewed by Mike Wyrzykowski.

This patch re-organizes the Overlay Region management code, both for
performance reasons and in preparation of further changes.
It also moves the code from WKWebViewIOS to
RemoteScrollingCoordinatorProxyIOS.

First, we remove the code responsible for snapping, merging and
filtering Overlay Region rects. The system consuming these rects
already implements the same logic.

The ScrollView selection for Overlay Regions now only runs when a
ScrollingNode is added, removed or when relevant properties like the
TotalContentsSize change.

Finally, the tree traversal to find Overlay Region rects is also
optimized to stop early and limit the amount walking up the ancestor
chain.

Tests: updated the expectation files after disabling snapping, merging
and filtering from the old code. Confirmed that the changes on top of
that do not change the expectations.

* LayoutTests/overlay-region/big-edge-overlay-expected.txt: Removed.
* LayoutTests/overlay-region/big-edge-overlay-horizontal-expected.txt: Removed.
* LayoutTests/overlay-region/big-edge-overlay-horizontal.html: Removed.
* LayoutTests/overlay-region/big-edge-overlay.html: Removed.
* LayoutTests/overlay-region/snapping-expected.txt: Removed.
* LayoutTests/overlay-region/snapping.html: Removed.
* LayoutTests/overlay-region/merging-expected.txt: Removed.
* LayoutTests/overlay-region/merging.html: Removed.
Remove tests that are no longer relevant.

* LayoutTests/overlay-region/full-page-dynamic-expected.txt:
* LayoutTests/overlay-region/full-page-expected.txt:
* LayoutTests/overlay-region/full-page-overflow-expected.txt:
* LayoutTests/overlay-region/full-page-overflow-scrolling-expected.txt:
* LayoutTests/overlay-region/full-page-scrolling-expected.txt:
* LayoutTests/overlay-region/overlay-background-expected.txt:
Update expectations with snapping, merging and filtering off.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:mainFrameData:pageData:transactionID:]):
(addOverlayEventRegions): Deleted.
(snapRectToScrollViewEdges): Deleted.
(configureScrollViewWithOverlayRegionsIDs): Deleted.
(-[WKWebView _scrollViewCanHaveOverlayRegions:]): Deleted.
(-[WKWebView _selectOverlayRegionScrollView:]): Deleted.
(-[WKWebView _updateOverlayRegions]): Deleted.
(-[WKWebView _resetOverlayRegions]): Deleted.
(-[WKWebView _updateScrollCoordinatorProxyForOverlayRegions:]): Deleted.
Remove the code from WKWebViewIOS.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
Update Overlay Regions as part of the layer tree transaction commit.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::updateOverlayRegions):
(WebKit::RemoteScrollingCoordinatorProxy::overlayRegionsEnabledChanged):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateOverlayRegions):
(WebKit::RemoteScrollingCoordinatorProxyIOS::overlayRegionsEnabledChanged):
(WebKit::RemoteScrollingCoordinatorProxyIOS::selectOverlayRegionScrollViewIfNeeded):
(WebKit::overlayDrawsAboveScrollView):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateOverlayRegionLayers):
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
(WebKit::RemoteScrollingCoordinatorProxyIOS::establishLayerTreeScrollingRelations):
(WebKit::RemoteScrollingCoordinatorProxyIOS::removeDestroyedLayerIDs): Deleted.
(WebKit::RemoteScrollingCoordinatorProxyIOS::fixedScrollingNodeLayerIDs const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxyIOS::overlayRegionCandidates const): Deleted.
Add the improved logic back to the RemoteScrollingCoordinatorProxyIOS.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::preferencesDidChange):
Signal the ScrollingCoordinator, and not the WebView, when the
preference changes.

* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView setScrollEnabled:]):
(-[WKScrollView _setScrollEnabledInternal:]):
Avoid calling `_updateScrollability` when the value doesn&apos;t change since
it triggers non-trivial work downstream for Overlay Regions.

Canonical link: <a href="https://commits.webkit.org/303903@main">https://commits.webkit.org/303903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3197706db0282d2b491e5f8e61c0de959f2c935e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133732 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6242 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44913 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141307 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85791 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6105 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102299 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69639 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136679 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4808 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119881 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83102 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4688 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2269 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113839 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37998 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143953 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5911 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38580 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110678 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5995 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5050 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110868 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28161 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4506 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116137 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59658 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5964 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34438 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5810 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69428 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6055 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5918 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebKit/UIProcess/ios/WKScrollView.mm:
```diff
@@ -455,12 +455,18 @@ - (BOOL)isScrollEnabled
 
 - (void)setScrollEnabled:(BOOL)value
 {
+    if (value == _scrollEnabledByClient)
```
This seems fine but I wonder if there is any situation where redundantly calling `-[UIScrollView setScrollEnabled:]` results in different behavior

etiennesegonzac replied:
not one that's captured by a test at least :)

Although we could gate the early returns with a 
`#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)` just to be safe?

mwyrzykowski replied:
I think it's fine as you have it 👍 - I can't convince myself looking at `setScrollEnabled:` in UIScrollView that this could possibly cause problems

mwyrzykowski reviewed (approved):
~~One question but otherwise~~ ✅

etiennesegonzac committed:
[visionOS] Update OverlayRegion configuration more deliberately
https://bugs.webkit.org/show_bug.cgi?id=303394
<rdar://163802576>

Reviewed by Mike Wyrzykowski.

This patch re-organizes the Overlay Region management code, both for
performance reasons and in preparation of further changes.
It also moves the code from WKWebViewIOS to
RemoteScrollingCoordinatorProxyIOS.

First, we remove the code responsible for snapping, merging and
filtering Overlay Region rects. The system consuming these rects
already implements the same logic.

The ScrollView selection for Overlay Regions now only runs when a
ScrollingNode is added, removed or when relevant properties like the
TotalContentsSize change.

Finally, the tree traversal to find Overlay Region rects is also
optimized to stop early and limit the amount walking up the ancestor
chain.

Tests: updated the expectation files after disabling snapping, merging
and filtering from the old code. Confirmed that the changes on top of
that do not change the expectations.

* LayoutTests/overlay-region/big-edge-overlay-expected.txt: Removed.
* LayoutTests/overlay-region/big-edge-overlay-horizontal-expected.txt: Removed.
* LayoutTests/overlay-region/big-edge-overlay-horizontal.html: Removed.
* LayoutTests/overlay-region/big-edge-overlay.html: Removed.
* LayoutTests/overlay-region/snapping-expected.txt: Removed.
* LayoutTests/overlay-region/snapping.html: Removed.
* LayoutTests/overlay-region/merging-expected.txt: Removed.
* LayoutTests/overlay-region/merging.html: Removed.
Remove tests that are no longer relevant.

* LayoutTests/overlay-region/full-page-dynamic-expected.txt:
* LayoutTests/overlay-region/full-page-expected.txt:
* LayoutTests/overlay-region/full-page-overflow-expected.txt:
* LayoutTests/overlay-region/full-page-overflow-scrolling-expected.txt:
* LayoutTests/overlay-region/full-page-scrolling-expected.txt:
* LayoutTests/overlay-region/overlay-background-expected.txt:
Update expectations with snapping, merging and filtering off.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:mainFrameData:pageData:transactionID:]):
(addOverlayEventRegions): Deleted.
(snapRectToScrollViewEdges): Deleted.
(configureScrollViewWithOverlayRegionsIDs): Deleted.
(-[WKWebView _scrollViewCanHaveOverlayRegions:]): Deleted.
(-[WKWebView _selectOverlayRegionScrollView:]): Deleted.
(-[WKWebView _updateOverlayRegions]): Deleted.
(-[WKWebView _resetOverlayRegions]): Deleted.
(-[WKWebView _updateScrollCoordinatorProxyForOverlayRegions:]): Deleted.
Remove the code from WKWebViewIOS.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
Update Overlay Regions as part of the layer tree transaction commit.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::updateOverlayRegions):
(WebKit::RemoteScrollingCoordinatorProxy::overlayRegionsEnabledChanged):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateOverlayRegions):
(WebKit::RemoteScrollingCoordinatorProxyIOS::overlayRegionsEnabledChanged):
(WebKit::RemoteScrollingCoordinatorProxyIOS::selectOverlayRegionScrollViewIfNeeded):
(WebKit::overlayDrawsAboveScrollView):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateOverlayRegionLayers):
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
(WebKit::RemoteScrollingCoordinatorProxyIOS::establishLayerTreeScrollingRelations):
(WebKit::RemoteScrollingCoordinatorProxyIOS::removeDestroyedLayerIDs): Deleted.
(WebKit::RemoteScrollingCoordinatorProxyIOS::fixedScrollingNodeLayerIDs const): Deleted.
(WebKit::RemoteScrollingCoordinatorProxyIOS::overlayRegionCandidates const): Deleted.
Add the improved logic back to the RemoteScrollingCoordinatorProxyIOS.

* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::preferencesDidChange):
Signal the ScrollingCoordinator, and not the WebView, when the
preference changes.

* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView setScrollEnabled:]):
(-[WKScrollView _setScrollEnabledInternal:]):
Avoid calling `_updateScrollability` when the value doesn't change since
it triggers non-trivial work downstream for Overlay Regions.

Canonical link: https://commits.webkit.org/303903@main
